### ISSUE TITLE
fit function

### ISSUE THREAD

**Issue by rojinsafavi** [skorch-dev/skorch#492]
I am using pytorch, and my data is X(input) and Y(label) and Z. Z is only required for the loss calculation and for gridseachCV score matrix calculation(I have built my own). I noticed that .fit only takes X and Y, is there a way to include Z too?

**Comment by ottonemo** [comment]
Can you show us what you have done so far so we can better answer your question?

**Comment by rojinsafavi** [comment]
I am using a very specific loss function and accuracy score which needs X, Y, and Z. X is the input feature matrix( shape NxM), Y is the label that we want to predict, and Z is has a form of [0,1,0,0,0,1...,0] shape N, so for each sample, I have one feature array, one label and 1 value for Z ( o or 1)

**Comment by BenjaminBossan** [comment]
Would it work if you put the `Z` either inside of `X` or inside of `Y`. E.g. your `Y` and `Z` are probably a shape `(n,)` arrays, you could concatenate them to a `(n, 2)` array. You probably already have a custom loss function, you would only need to separate the array again there.

**Comment by ottonemo** [comment]
As Benjamin said you have the option to either include your `z` into `X` or move it to `y`, either as a dict or an array with two dimensions `(y.shape == (num_samples, 2))`.

Here's are two examples:

## Putting z into X

```python
class MyModule(torch.nn.Module):
    def __init__(self, **kwargs):
        super().__init__()
        self.net = torch.nn.Sequential(
            torch.nn.Linear(20, 2),
            torch.nn.Softmax(dim=-1),
        )
    def forward(self, X, z=None):  # module gets z parameter as well since it is in X
        return self.net(X)

class MyClassifier(skorch.NeuralNetClassifier):        
    def get_loss(self, y_pred, y_true, X, training=False):
        z = X['z'].byte()
        return self.criterion_(y_pred[z], y_true[z])

def my_score(net, ds, y):
    Xz, y = skorch.utils.data_from_dataset(ds)
    z = Xz['z']
    y_pred = net.predict(X)
    return sklearn.metrics.accuracy_score(y[z], y_pred[z])

net = MyClassifier(
    MyModule, 
    callbacks=[
        # scoring=None => use net.score()
        skorch.callbacks.EpochScoring(scoring=my_score),
    ]
)

X, y = make_classification(n_samples=1000, n_features=20, n_classes=2)
X = X.astype(np.float32)
z = np.random.binomial(1, 0.5, size=y.shape)
z = z.astype(np.float32)

XZ = dict(X=X, z=z)
net.fit(XZ, y)
```

## Putting z into y

```python
class MaskedCVSplit(skorch.dataset.CVSplit):
    def check_cv(self, y):
        return super().check_cv(y['y'][y['mask']])  # extract y to compute train/valid split

class MaskedNLLLoss(torch.nn.NLLLoss):        
    def forward(self, y_pred, yz):
        y = yz['y']
        m = yz['mask']
        return super().forward(y_pred[m], y[m])

def masked_scoring(net, ds, y_):
    X, yz = skorch.utils.data_from_dataset(ds)
    y = yz['y']
    m = yz['mask']
    y_pred = net.predict(ds)
    return sklearn.metrics.accuracy_score(y[m], y_pred[m])

def y_extractor(y):
    return y['y']

net = skorch.NeuralNetClassifier(
    module, 
    train_split=MaskedCVSplit(), 
    criterion=MaskedNLLLoss,
    
    callbacks__valid_acc__scoring=masked_scoring,
    callbacks__valid_acc__target_extractor=y_extractor,
)

X, y = make_classification(n_samples=1000, n_features=20, n_classes=2)
X = X.astype(np.float32)
z = np.random.binomial(1, 0.5, size=y.shape)
z = z.astype(np.float32)

yz = dict(y=y, mask=z)

net.fit(X, yz)
```

**Comment by BenjaminBossan** [comment]
@rojinsafavi Did this solve your issue?

**Comment by rojinsafavi** [comment]
Thanks for your suggestion, I will try that I think it should be fine. I'll let you know if it does not work properly :-)

**Comment by rojinsafavi** [comment]
I decided to go with the first example. I have one more question though, I would like to do gridsearchCV with skorch to find the best hyperparameters for my model. And I am doing a regression prediction.  For the gridsearch, I would like to use my own scoring metric I know that this is done with "make_scorer" in sklearn. Is that the same in skorch?

**Comment by BenjaminBossan** [comment]
Great, I hope it works.

Regarding the scoring, if you use the normal sklearn `GridSearchCV`, your scoring function should work the same with skorch as with any other sklearn model.

Please feel free to close the issue once everything is resolved.

**Comment by ottonemo** [comment]
This seems to be resolved. Feel free to re-open or open a new issue if this should not be the case.
